data_juicer/ops/mapper/image_diffusion_mapper.py

```diff
@@ -0,0 +1,205 @@
+        :param caption_key: the key name of fields in samples to store captions
+            for each images. It can be a string if there is only one image in
+            each sample. Otherwise, it should be a list. If it's none,
+            ImageDiffusionMapper will produce captions for each images.
+        :param hf_blip2: blip2 model name on huggingface to generate caption if
+            caption_key is None.
+        """
+        super().__init__(*args, **kwargs)
+        self._batched_op = True
+        self.strength = strength
+        self.guidance_scale = guidance_scale
+        self.aug_num = aug_num
+        self.keep_original_sample = keep_original_sample
+        self.caption_key = caption_key
+        self.prompt = 'A photo of a '
+        if not self.caption_key:
+            from .generate_caption_mapper import GenerateCaptionMapper
+            self.op_generate_caption = GenerateCaptionMapper(
+                hf_blip2=hf_blip2,
+                keep_original_sample=False,
+                prompt=self.prompt)
+
+        self.model_key = prepare_model(model_type='diffusion',
+                                       model_name_or_path=hf_diffusion)
+        self.diffusion_model = get_model(model_key=self.model_key)
+
+    def _real_guidance(self, caption: str, image: Image.Image):
+
+        canvas = image.resize((512, 512), Image.BILINEAR)
+        prompt = caption
+
+        kwargs = dict(image=canvas,
+                      prompt=[prompt],
+                      strength=self.strength,
+                      guidance_scale=self.guidance_scale)
+
+        has_nsfw_concept = True
+        while has_nsfw_concept:
+            outputs = self.diffusion_model(**kwargs)
+
+            has_nsfw_concept = (self.diffusion_model.safety_checker is not None
+                                and outputs.nsfw_content_detected[0])
+
+        canvas = outputs.images[0].resize(image.size, Image.BILINEAR)
+
+        return canvas
+
+    def _process_single_sample(self, ori_sample, context=False):
+        """
+        :param ori_sample: a single data sample before applying generation
+        :return: batched results after generation
+        """
+        # there is no image in this sample
+        if self.image_key not in ori_sample or \
+                not ori_sample[self.image_key]:
+            return []
+
+        # load images
+        loaded_image_keys = ori_sample[self.image_key]
+        ori_sample, images = load_data_with_context(ori_sample, context,
+                                                    loaded_image_keys,
+                                                    load_image)
+
+        # load captions
+        if self.caption_key:
+            captions = ori_sample[self.caption_key]
+            if not isinstance(captions, list):
+                captions = [captions]
+            assert len(captions) == len(
+                images), 'The num of captions must match the num of images.'
```

Comment:
For some datasets (e.g. mmc4), there might be multiple images corresponding to one sentence or caption. Can we find a more proper way to handle this situation?

Reply:
See the submission [one caption all images](https://github.com/alibaba/data-juicer/pull/200/commits/038c420dfe1667424945a7f98d0afe58a8651cee).